### PR TITLE
Embed Dart DevTools in Chrome DevTools 

### PR DESCRIPTION
**Pull request by elliette**
* Adds a Dart panel to Chrome DevTools with the Dart DevTools debugger view embedded 
* Currently the Dart panel is only added for Dart apps with the `data-ddr-dart-app`, this is to gate the feature for internal Dart apps for now (currently they are the only ones that have the correct response headers for IFRAME-ing). 
* Also provides a quick fix for https://github.com/dart-lang/webdev/issues/1501

### CODE REVIEW

File: dwds/debug_extension/CONTRIBUTING.md [hunk 1]
```diff
@@ -6,6 +6,7 @@
 pub run build_runner build web -o build -r
 ```
 
+TODO(elliette): Building with DDC no longer works, figure out why.
```
Comment (grouma) [inline dwds/debug_extension/CONTRIBUTING.md:9]: Are you loading the extension correctly from the built directory?

File: dwds/debug_extension/web/background.dart [hunk 2]
```diff
@@ -123,23 +157,37 @@ void main() {
...
   queryTabs(getCurrentTabQuery, allowInterop((List<Tab> tabs) {
-    attachDebuggerToTab(tabs[0]);
+    if (tabs != null && tabs.isNotEmpty) {
+      attachDebuggerToTab(tabs[0]);
+    } else if (_mostRecentDartTab != null) {
+      attachDebuggerToTab(_mostRecentDartTab);
+    } else {
+      alert('''
+          Could not find a Dart app to start debugging. 
+          Has the Dart Debug Extension icon turned blue yet?
```
Comment (grouma) [inline dwds/debug_extension/web/background.dart:188]: Maybe rephrase this to: `The Dart Debug Extension will turn blue when a Dart application is detected.`

File: dwds/debug_extension/web/manifest.json [hunk 3]
```diff
@@ -10,22 +10,22 @@
...
+    "content_scripts": [{
+        "matches": [
+            "<all_urls>"
+        ],
+        "js": [
+            "detector.js"
+        ],
+        "run_at": "document_end"
+    }]
+}
```
Comment (grouma) [inline dwds/debug_extension/web/manifest.json:31]: [nit] newline

File: dwds/debug_extension/web/panel.html [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+<body>
+
+  <div class='flex-row'>
+    <div class='flex-column'>
+      <button class="material-button" id="debuggingButton" type="button">Start debugging</button>
+      <h1>Please read:</h1>
+      <p>Before debugging, please <strong>disable</strong> focus on the Sources panel for breakpoints:</p>
+      <code>Settings > Preferences > Sources > uncheck "Focus Sources panel when triggering a breakpoint"</code>
+      <img src="devtools_settings.png" alt="Picture of the setting in Chrome DevTools to disable.">
+      <p>And make sure the Dart Debug Extension is <strong>pinned</strong> to your Chrome profile:</p>
```
Comment (grouma) [inline dwds/debug_extension/web/panel.html:17]: Why must they pin it?

**Review by grouma — COMMENTED** [review]
What happens in the following scenarios:

1) You have chrome devtools opened and connected but you try to click the extension again?
2) You click `Start Debugging` before a Dart app is detected?

File: dwds/debug_extension/CONTRIBUTING.md (hunk 1 above)
Comment (elliette) [reply dwds/debug_extension/CONTRIBUTING.md:9]: To clarify, there are two directories:
1. `/dwds/debug_extension/web`
2. `/dwds/debug_extension/build/web`
Which one is the correct one to load from? Is it (1) for dart2js and (2) for DDC?

File: dwds/debug_extension/web/background.dart (hunk 2 above)
Comment (elliette) [reply dwds/debug_extension/web/background.dart]: That's better, changed to that

File: dwds/debug_extension/web/manifest.json (hunk 3 above)
Comment (elliette) [reply dwds/debug_extension/web/manifest.json:31]: This is how it autoformats on save for me 🤷‍♀️

File: dwds/debug_extension/web/panel.html (hunk 4 above)
Comment (elliette) [reply dwds/debug_extension/web/panel.html]: Ah good point, they don't have to for this. Removed that part of the instructions.

File: dwds/debug_extension/CONTRIBUTING.md (hunk 1 above)
Comment (grouma) [reply dwds/debug_extension/CONTRIBUTING.md:9]: That's correct, unless something has changed.

File: dwds/debug_extension/web/manifest.json (hunk 3 above)
Comment (grouma) [reply dwds/debug_extension/web/manifest.json:31]: You'll want to update your IDE settings.

Comment (elliette) [reply dwds/debug_extension/web/manifest.json]: Got it working, added old formatting back. Thanks!

File: dwds/debug_extension/CONTRIBUTING.md (hunk 1 above)
Comment (elliette) [reply dwds/debug_extension/CONTRIBUTING.md:9]: Interesting. I've opened https://github.com/dart-lang/webdev/issues/1506 to investigate (this occurs on `master` branch as well)

**Comment by elliette** [comment]
> What happens in the following scenarios:
> 
> 1. You have chrome devtools opened and connected but you try to click the extension again?
> 2. You click `Start Debugging` before a Dart app is detected?

1. If you click the Dart Debug Extension while the embed Dart DevTools is debugging the app, you will see an alert: "DevTools is already opened on a different window." We could change that alert to make it more generic since it's not *technically* a different window.
2. If you click the button before a Dart app is detected, nothing will happen. Again, we could add an alert or error message so that the user knows they should wait. 

Wdyt? These could be in this PR or follow up polish items.

**Comment by grouma** [comment]
> > What happens in the following scenarios:
> > 
> > 1. You have chrome devtools opened and connected but you try to click the extension again?
> > 2. You click `Start Debugging` before a Dart app is detected?
> 
> 1. If you click the Dart Debug Extension while the embed Dart DevTools is debugging the app, you will see an alert: "DevTools is already opened on a different window." We could change that alert to make it more generic since it's not _technically_ a different window.

Updating so the message is more generic makes sense to me.
> 2. If you click the button before a Dart app is detected, nothing will happen. Again, we could add an alert or error message so that the user knows they should wait.
> 
Can we not make the button conditionally enabled?

> Wdyt? These could be in this PR or follow up polish items.

It's fine to be left for a follow up PR.

File: dwds/debug_extension/CONTRIBUTING.md (hunk 1 above)
Comment (grouma) [reply dwds/debug_extension/CONTRIBUTING.md:9]: Update the TODO with the issue number please.

**Review by grouma — APPROVED** [review]
